### PR TITLE
Add support for NFS and SMB/CIFS network file systems

### PR DESCRIPTION
**Pull request by agners**
These user space helpers are required to mount NFS or SMB/CIFS network file systems.

### CODE REVIEW

**Comment by jens-maus** [comment]
> These user space helpers are required to mount NFS or SMB/CIFS network file systems.

Is this really the case? I thought these util packages are only required when running a SMB/CIFS and/or NFS server and are not required for a client? At least in case of RaspberryMatic these package seem to be not required for mounting NFS/CIFS shares in any way.

**Comment by agners** [comment]
Yes, this is clearly the case. Trying to mount CIFS without this leads to:

```
mount: /mnt/data/smbtest: mount(2) system call failed: No route to host.
       dmesg(1) may have more information after failed mount system call.
```
and the following in `dmesg`
```
CIFS: Unable to determine destination address
```

Trying to mount NFS without utilities leads to:
```
mount: /mnt/data/nfstest2: bad option; for several filesystems (e.g. nfs, cifs) you might need a /sbin/mount.<type> helper program.
       dmesg(1) may have more information after failed mount system call.
```

With what command do you mount NFS/CIFS?

**Comment by jens-maus** [comment]
> With what command do you mount NFS/CIFS?

We simply mount CIFS/NFS via the `mount` commands and have documented it here:

https://github.com/jens-maus/RaspberryMatic/wiki/Experten-Features#anbinden-eines-externen-nas-nfscifs-netzwerk-laufwerkes

So for a CIFS/SMB mount on a NAS we have documented it to look like:

```
mount -t cifs -o noserverino,vers=3.11,username=USER,password=PASS //192.168.1.1/fritz.nas/test/RM_B /mnt
```

And for NFS I personally use:

```
mount -t nfs -o nolock,hard,intr,relatime,fsc 192.168.X.X:/volume1/Backup/share /mnt
``` 

And all this works without these CIFS/NFS utilities packages from buildroot. Strange that HAos seem to require them?!? I would be curious to find out why this seems to be required for HAos while for the similar buildroot-based RaspberryMatic it is obviously not required?!?

**Comment by agners** [comment]
Uh I see, at least CIFS works with IP only. But honestly, host name support is kinda a must in 2023 :wink: 

NFS however seems to require the utils when using IP or host name.

**Comment by jens-maus** [comment]
> > And all this works without these CIFS/NFS utilities packages from buildroot. Strange that HAos seem to require them?!?
> 
> That is indeed strange. I'd be happy if they are not needed. Can you double check they are not present?
> 
> ```
> ls -la /usr/sbin/mount.*
> ```

This is the result of the `ls`:

```
:~# ls -la /usr/sbin/mount.*
-rwxr-xr-x    1 root     root         18672 Apr  3 04:12 /usr/sbin/mount.fuse3
```

So indeed, the `mount.nfs` or `mount.cifs` is not present here.

**Comment by jens-maus** [comment]
> NFS however seems to require the utils when using IP or host name.

Have you check all relevant CIFS/NFS kernel options? At least here

```
# CONFIG_NFS_USE_LEGACY_DNS is not set
CONFIG_NFS_USE_KERNEL_DNS=y
```

This could be related to this matter? See here for a full kernel defconfig we are using for RaspberryMatic which could reveal some kernel feature differences that might explain that all this works without these cifs/nfs utils:

https://github.com/jens-maus/RaspberryMatic/blob/master/buildroot-external/board/ova/kernel_defconfig

Or this could be related to implicit busybox features not activated in your case? See here:

https://github.com/jens-maus/RaspberryMatic/blob/master/buildroot-external/Busybox.config#L671

**Comment by agners** [comment]
> Have you check all relevant CIFS/NFS kernel options? At least here
> 
> ```
> # CONFIG_NFS_USE_LEGACY_DNS is not set
> CONFIG_NFS_USE_KERNEL_DNS=y
> ```

Same here, so this seems not to be it.

Can you mount CIFS with hostname?

All documentation I read they usually install those utils first before even attempting to mount anything. I guess there is also a reason that they even exist in first place? I'd guess that some options are parsed/processed in user space, maybe. So not using them might lead to a compromised NFS/CIFS feature set.

**Comment by agners** [comment]
> Or this could be related to implicit busybox features not activated in your case? See here:
> 
> https://github.com/jens-maus/RaspberryMatic/blob/master/buildroot-external/Busybox.config#L671

Oh I see that could be it. HAOS is using `mount` from util-linux

```
# mount --version
mount from util-linux 2.38 (libmount 2.38.0: btrfs, namespaces, assert, debug)
```

**Comment by jens-maus** [comment]
> Oh I see that could be it. HAOS is using `mount` from util-linux
> 
> ```
> # mount --version
> mount from util-linux 2.38 (libmount 2.38.0: btrfs, namespaces, assert, debug)
> ```

Yes, that's probably the reason for the difference. We are using mount from busybox with the shown extensions. However, I dunno what might be the best / optimal approach here (util-linux + nfs/cifs-utils or simple busybox)...

**Comment by agners** [comment]
I actually realized that some server component got enabled by default as well. Opted out of that, but enabled NFSv4. support. This makes the overhead even a bit less (<400KiB).

File: buildroot-external/configs/generic_aarch64_defconfig [hunk 1]
```diff
@@ -87,7 +87,8 @@ BR2_PACKAGE_DROPBEAR=y
 # BR2_PACKAGE_IFUPDOWN_SCRIPTS is not set
 BR2_PACKAGE_NETWORK_MANAGER=y
 BR2_PACKAGE_NETWORK_MANAGER_CLI=y
-BR2_PACKAGE_NFS_UTILS=y
+BR2_PACKAGE_NFS_UTILS=yBR2_PACKAGE_NFS_UTILS_NFSV4=y
```
[the remark this answers is not in the document]
Comment (agners) [reply buildroot-external/configs/generic_aarch64_defconfig]: And with NFSv4 correctly enabled, this change now require 1.75MiB :cold_face:

[the remark this answers is not in the document]
Comment (agners) [reply buildroot-external/configs/generic_aarch64_defconfig]: Hm, it seems NFSv4 support is compiled into mount unconditionally:

https://github.com/stefanha/nfs-utils/blob/708a963f5eb1892e39db1470d46429d8d8e17e38/utils/mount/Makefile.am#L12-L14


In a quick test the `BR2_PACKAGE_NFS_UTILS_NFSV4` was not required to mount a NFSv4 exported share (Wireshark reports NFSv4 packages). So I'll drop that config option for now.

**Comment by leokeba** [comment]
Hello, I have been investigating an issue that is at least tangentially related to this one (cf previous mention), and I've discovered that there seems to be a kernel module missing in the aarch64 distribution preventing the use of utf-8 encoding when mounting cifs shares on haos.

I am willing to dig a bit more, but my linux sysadmin skills are a bit overloaded and I don't really know how to proceed, does anyone have a suggestion on how to help it being resolved ?

**Comment by agners** [comment]
> Hello, I have been investigating an issue that is at least tangentially related to this one (cf previous mention), and I've discovered that there seems to be a kernel module missing in the aarch64 distribution preventing the use of utf-8 encoding when mounting cifs shares on haos.

Hm, it seems that generic-aarch64 is using `CONFIG_NLS_DEFAULT="iso8859-1"`, whereas other kernel configs are using `CONFIG_NLS_DEFAULT="utf8"`. This should be relatively easy to fix. I guess nowadays a global default of UTF-8 makes sense.

**Comment by leokeba** [comment]
Should I post this as an issue ?

**Comment by agners** [comment]
Not necessary, I already created a PR: #2450.

**Comment by leokeba** [comment]
Thanks that was quick, and sorry for randomly posting here by the way, it seemed related enough so I didn't really think twice.

**Pull request merged**